Mptcp v1

CrapsDorian opened the pull request:
Multipath TCP (MPTCP), standardized in RFC8684 [1], is a TCP extension
that enables a TCP connection to use different paths.

Multipath TCP has been used for several use cases. On smartphones, MPTCP
enables seamless handovers between cellular and Wi-Fi networks while
preserving established connections. This use-case is what pushed Apple
to use MPTCP since 2013 in multiple applications [2]. On dual-stack
hosts, Multipath TCP enables the TCP connection to automatically use the
best performing path, either IPv4 or IPv6. If one path fails, MPTCP
automatically uses the other path.

To benefit from MPTCP, both the client and the server have to support
it. Multipath TCP is a backward-compatible TCP extension that is enabled
by default on recent Linux distributions (Debian, Ubuntu, Redhat, ...).
Multipath TCP is included in the Linux kernel since version 5.6 [3]. To
use it on Linux, an application must explicitly enable it when creating
the socket. No need to change anything else in the application.

This attached patch adds an mptcp option in the config which allows the 
creation of an MPTCP socket instead of TCP on Linux. If Multipath TCP 
is not supported on the system, an error will be reported and the start failed. 

Due to the limited impact within a data center environment, 
we have decided not to implement MPTCP between the proxy and the servers. 
The high-speed, low-latency nature of data center networks reduces 
the benefits of MPTCP, making the complexity of its implementation 
unnecessary in this context.

Developed with the help of Matthieu Baerts (matttbe@kernel.org) and
Olivier Bonaventure (Olivier.Bonaventure@uclouvain.be)

Link: https://www.rfc-editor.org/rfc/rfc8684.html [1]
Link: https://www.tessares.net/apples-mptcp-story-so-far/ [2]
Link: https://www.mptcp.dev [3]

matttbe commented on include/haproxy/global-t.h:
```diff
@@ -85,6 +85,7 @@
 #define GTUNE_LISTENER_MQ_OPT    (1<<28)
 #define GTUNE_LISTENER_MQ_ANY    (GTUNE_LISTENER_MQ_FAIR | GTUNE_LISTENER_MQ_OPT)
 #define GTUNE_QUIC_CC_HYSTART    (1<<29)
+#define GTUNE_MPTCP      		 (1<<30)
```
the alignment is not OK because you use tabs apparently

matttbe replied:
Mmh, I just noticed you fixed that in the 3rd commit. Please next time squash such commits, and do 'rebase' on top of master, no 'merge' with master which creates an extra commit we don't need to review.

matttbe commented on include/haproxy/global-t.h:
```diff
@@ -85,6 +85,7 @@
 #define GTUNE_LISTENER_MQ_OPT    (1<<28)
 #define GTUNE_LISTENER_MQ_ANY    (GTUNE_LISTENER_MQ_FAIR | GTUNE_LISTENER_MQ_OPT)
 #define GTUNE_QUIC_CC_HYSTART    (1<<29)
+#define GTUNE_MPTCP      		 (1<<30)
```
Should we use `GTUNE_NO_MPTCP`, similar to `GTUNE_NO_QUIC`?

EDIT: but then if you do that, you should support a fallback in case of error when creating the socket: re-create a socket with the protocol set to 0

matttbe commented on src/cfgparse-global.c:
```diff
@@ -1334,6 +1334,11 @@ int cfg_parse_global(const char *file, int linenum, char **args, int kwm)
+        if (alertif_too_many_args(0, file, linenum, args, &err_code))
+            goto out;
+		global.tune.options |= GTUNE_MPTCP;
+    }
```
the alignment is not OK, you used a mix of tabs and spaces

matttbe commented on src/listener.c:
```diff
@@ -399,6 +399,11 @@ void stop_listener(struct listener *l, int lpx, int lpr, int lli)
  */
 void default_add_listener(struct protocol *proto, struct listener *listener)
 {
+	if(global.tune.options & GTUNE_MPTCP)
```
please respect the code style: space after `if`

matttbe commented on src/listener.c:
```diff
@@ -399,6 +399,11 @@ void stop_listener(struct listener *l, int lpx, int lpr, int lli)
 {
+	if(global.tune.options & GTUNE_MPTCP)
+		/* check MPTCP option */
+#ifdef linux
```
if you are not on Linux, the code will be

```c
if (global.tune.options & GTUNE_MPTCP)
  if (listener->state != LI_INIT)
    return;
```

You need to add the `ifdef` above the `if`

matttbe replied:
are you sure it is not `#ifdef __linux__`?

matttbe commented on src/listener.c:
```diff
@@ -399,6 +399,11 @@ void stop_listener(struct listener *l, int lpx, int lpr, int lli)
 void default_add_listener(struct protocol *proto, struct listener *listener)
 {
+	if(global.tune.options & GTUNE_MPTCP)
+		/* check MPTCP option */
```
check? Use?

matttbe commented on doc/haproxy.1:
```diff
@@ -1,4 +1,4 @@
-.TH HAPROXY 1 "17 August 2007" 
+.TH HAPROXY 1 "17 August 2007"
```
please undo this

matttbe commented on doc/configuration.txt:
```diff
@@ -2958,6 +2959,10 @@ maxzlibmem <number>
 
+mptcp
+  Enables the use of the Multipath TCP (MPTCP) protocol. This option is only
+  available when support for MPTCP was built in.
```
what do you mean with "MPTCP was built in"?

if the kernel supports it? Currently limited to Linux?

matttbe commented on src/listener.c:
```diff
@@ -399,6 +399,11 @@ void stop_listener(struct listener *l, int lpx, int lpr, int lli)
+	if(global.tune.options & GTUNE_MPTCP)
+		/* check MPTCP option */
+#ifdef linux
+		proto->sock_prot = IPPROTO_MPTCP;
```
You will certainly have to define `IPPROTO_MPTCP` if it is not defined already

matttbe commented on src/cfgparse-global.c:
```diff
@@ -1334,6 +1334,11 @@ int cfg_parse_global(const char *file, int linenum, char **args, int kwm)
 			HA_ATOMIC_STORE(&global.anon_key, tmp);
 		}
 	}
+	else if (strcmp(args[0], "mptcp") == 0) {
```
no need to modify `common_kw_list`?

matttbe reviewed (changes requested):
good start, maybe MPTCP should be used by default and a fallback to TCP should be done in case of error?

matttbe commented on src/cfgparse-global.c:
```diff
@@ -51,7 +51,7 @@ static const char *common_kw_list[] = {
 	"presetenv", "unsetenv", "resetenv", "strict-limits", "localpeer",
 	"numa-cpu-mapping", "defaults", "listen", "frontend", "backend",
-	"peers", "resolvers", "cluster-secret", "no-quic", "limited-quic",
+	"peers", "resolvers", "cluster-secret", "no-quic", "limited-quic", "no-mptcp",
```
80 chars max

matttbe commented on src/cfgparse-global.c:
```diff
@@ -51,8 +51,8 @@ static const char *common_kw_list[] = {
-	"peers", "resolvers", "cluster-secret", "no-quic", "limited-quic",
-	NULL /* must be last */
+	"peers", "resolvers", "cluster-secret", "no-quic", "limited-quic", 
+	"no-mptcp",	NULL /* must be last */
```
(extra tab after `no-mptcp`)

matttbe commented on src/cfgparse-global.c:
```diff
@@ -1334,6 +1334,11 @@ int cfg_parse_global(const char *file, int linenum, char **args, int kwm)
+        if (alertif_too_many_args(0, file, linenum, args, &err_code))
+            goto out;
+        global.tune.options |= GTUNE_NO_MPTCP;
+	}
```
mix of spaces and tabs: tabs should be used

matttbe commented on src/listener.c:
```diff
@@ -399,6 +399,18 @@ void stop_listener(struct listener *l, int lpx, int lpr, int lli)
+#define IPPROTO_MPTCP 262
+#endif
+        // proto->sock_prot = IPPROTO_MPTCP;
+	}
```
mix of spaces and tabs: tabs should be used

matttbe commented on doc/configuration.txt:
```diff
@@ -2958,6 +2959,9 @@ maxzlibmem <number>
   "ZlibMemUsage" in bytes.
 
+mptcp
+  uses MPTCP instead of TCP.
```
Capital letter: 'Uses'

While at it:

```
  Uses MPTCP instead of TCP. Multipath TCP or MPTCP is an extension to the
  standard TCP and is described in RFC 8684. It allows a device to make use of
  multiple interfaces at once to send and receive TCP packets over a single
  MPTCP connection. MPTCP can aggregate the bandwidth of multiple interfaces or
  prefer the one with the lowest latency, it also allows a fail-over if one path
  is down, and the traffic is seamlessly reinjected on other paths.
```

(from mptcp.dev)

matttbe commented on src/cfgparse-global.c:
```diff
@@ -1334,6 +1334,11 @@ int cfg_parse_global(const char *file, int linenum, char **args, int kwm)
+        if (alertif_too_many_args(0, file, linenum, args, &err_code))
+            goto out;
+        global.tune.options |= GTUNE_MPTCP;
+	}
```
mix of spaces and tabs: tabs should be used

matttbe commented on include/haproxy/protocol-t.h:
```diff
@@ -99,6 +99,7 @@ struct protocol {
 	enum proto_type proto_type;                     /* protocol type at the socket layer (PROTO_TYPE_*) */
 	int sock_type;					/* socket type, as passed to socket()     */
 	int sock_prot;					/* socket protocol, as passed to socket() */
+	int sock_prot_mptcp;				/* MPTCP socket protocol , as passed to socket() */
```
I don't think you need this

matttbe commented on src/proto_tcp.c:
```diff
@@ -81,6 +81,9 @@ struct protocol proto_tcpv4 = {
 	.sock_type      = SOCK_STREAM,
 	.sock_prot      = IPPROTO_TCP,
+#ifdef __linux__
+	.sock_prot_mptcp= IPPROTO_MPTCP,
```
I don't think you need this. You need to modify `sock_prot` if the option is set.

matttbe commented on src/proto_tcp.c:
```diff
@@ -127,6 +130,9 @@ struct protocol proto_tcpv6 = {
 	.sock_type      = SOCK_STREAM,
 	.sock_prot      = IPPROTO_TCP,
+#ifdef __linux__
+	.sock_prot_mptcp= IPPROTO_MPTCP,
```
same here

matttbe commented on src/protocol.c:
```diff
@@ -103,8 +103,11 @@ int protocol_supports_flag(struct protocol *proto, uint flag)
+		if (global.tune.options & GTUNE_MPTCP)
+			ret = _sock_supports_reuseport(proto->fam, proto->sock_type, proto->sock_prot_mptcp);
+		else 
+			ret = _sock_supports_reuseport(proto->fam, proto->sock_type, proto->sock_prot);		
```
I think this block should not be modified: sock_prot could be UDP here

matttbe commented on src/sock_inet.c:
```diff
@@ -327,8 +327,14 @@ int sock_inet_bind_receiver(struct receiver *rx, char **errmsg)
+		                 rx->proto->sock_type, rx->proto->sock_prot_mptcp);
+		else
+			fd = my_socketat(rx->settings->netns, rx->proto->fam->sock_domain,
+		                	 rx->proto->sock_type, rx->proto->sock_prot);
```
I think this block should not be modified: sock_prot could be UDP here

matttbe commented on include/haproxy/global-t.h:
```diff
@@ -26,6 +26,12 @@
 #include <haproxy/freq_ctr-t.h>
 
+#ifdef __linux__
+#ifndef IPPROTO_MPTCP
```
if you only use it in one .c file, maybe best to declare it there (`src/cfgparse.c`), just before using it.

matttbe commented on src/cfgparse.c:
```diff
@@ -205,6 +205,12 @@ int str2listener(char *str, struct proxy *curproxy, struct bind_conf *bind_conf,
 	} /* end while(next) */
+#ifdef __linux__
+	if (global.tune.options & GTUNE_MPTCP) {
+		if (proto->sock_prot == IPPROTO_TCP)
```
```c
if (A) {
  if (B)
    (...)
}
```

Can be:

```c
if (A && B)
  (...)
}
```

matttbe commented on src/cfgparse.c:
```diff
@@ -205,6 +205,12 @@ int str2listener(char *str, struct proxy *curproxy, struct bind_conf *bind_conf,
 			goto fail;
 		}
 	} /* end while(next) */
+#ifdef __linux__
```
(see above: move `IPPROTO_MPTCP` def here)

matttbe commented on src/cfgparse.c:
```diff
@@ -205,6 +205,12 @@ int str2listener(char *str, struct proxy *curproxy, struct bind_conf *bind_conf,
+#ifdef __linux__
+	if (global.tune.options & GTUNE_MPTCP) {
+		if (proto->sock_prot == IPPROTO_TCP)
+			proto->sock_prot = IPPROTO_MPTCP;
```
mmh, did you even verify the modification you did?

It looks like here, you are only modifying the last `proto` (there is a loop above), after having created the listening socket. The modification should be done after (or in `str2sa_range()`) and before calling `create_listeners()`, no?

Please add a test.

matttbe commented on src/cfgparse.c:
```diff
@@ -205,6 +205,13 @@ int str2listener(char *str, struct proxy *curproxy, struct bind_conf *bind_conf,
+#endif
+	if (global.tune.options & GTUNE_MPTCP && proto->sock_prot == IPPROTO_TCP)
+		proto->sock_prot = IPPROTO_MPTCP;
+#endif
```
I'm still not sure how it can work.

Can you move this to the line 202, just before `create_listeners()`? (with an empty line before and after this new block)

matttbe replied:
OK, sorry, I now took some time to dig in the code: `create_listeners()` doesn't create the listerner socket, it only adds it to list (see `default_add_listener()`). The socket is created later, in `bind()`. Here, only the last `proto` is modified: so if by chance, the last socket is a TCP one, it will work. And it works because `proto` will be one item of `__protocol_by_family` array.

In short, this should be moved into the loop.

mux99 commented on src/cfgparse.c:
```diff
@@ -205,6 +205,13 @@ int str2listener(char *str, struct proxy *curproxy, struct bind_conf *bind_conf,
 	} /* end while(next) */
+#ifdef __linux__
+#ifndef IPPROTO_MPTCP
+#define IPPROTO_MPTCP
```
the protocol number seems to be missing (262)

matttbe commented on src/cfgparse-global.c:
```diff
@@ -1334,6 +1334,11 @@ int cfg_parse_global(const char *file, int linenum, char **args, int kwm)
+	else if (strcmp(args[0], "mptcp") == 0) {
+		if (alertif_too_many_args(0, file, linenum, args, &err_code))
+			goto out;
+		global.tune.options |= GTUNE_MPTCP;
```
I have another idea: what if you remove `GTUNE_MPTCP` (so here, in `global-t.h` and `cfgparse.c`), and instead, you add here:

```c
	else if (strcmp(args[0], "mptcp") == 0) {
		if (alertif_too_many_args(0, file, linenum, args, &err_code))
			goto out;
#ifdef __linux__
#ifndef IPPROTO_MPTCP
#define IPPROTO_MPTCP 262
#endif
		proto_tcpv4.sock_prot = IPPROTO_MPTCP;
		proto_tcpv6.sock_prot = IPPROTO_MPTCP;
		proto_rhttp.sock_prot = IPPROTO_MPTCP;
#else
		ha_alert("parsing [%s:%d]: '%s' is only supported on Linux.\n",
			 file, linenum, args[0]);
		err_code |= ERR_ALERT | ERR_FATAL;
		goto out;
#endif
	}
```

and in `include/haproxy/proto_rhttp.h`, adding this line after the `#include`

```c
extern struct protocol proto_rhttp;
```

By doing that, you modify the TCP (v4/v6) and the `rhttp` protocols to use `IPPROTO_MPTCP`.

I guess you will also need to modify this line: https://github.com/CrapsDorian/haproxy/blob/50d8c187423d6b7e9b1083e05370885f6d12e844/src/cfgparse.c#L1177 but I guess the CI will tell you that, e.g.:

```c
 if (proto->sock_prot != proto_tcpv4.sock_prot) {
```

matttbe reviewed (changes requested):
sorry for the misunderstanding, I now understand it can work (but by accident).

I have another idea to have a shorter code.

matttbe commented on src/cfgparse-global.c:
```diff
@@ -24,6 +24,11 @@
 #include <haproxy/tools.h>
+#include <haproxy/proto_rhttp.h>
+
+extern struct protocol proto_rhttp;
```
can you move it to `haproxy/proto_rhttp.h`, just under the `#include` instead please? So it would be similar to `proto_tcpv[46]` defined in `haproxy/proto_tcp.h`

matttbe replied:
> can you move it to `haproxy/proto_rhttp.h`, just under the `#include` instead please? So it would be similar to `proto_tcpv[46]` defined in `haproxy/proto_tcp.h`

do not forget this ↑

matttbe commented on src/cfgparse-global.c:
```diff
@@ -24,6 +24,11 @@
+
+extern struct protocol proto_rhttp;
+extern struct protocol proto_tcpv4;
+extern struct protocol proto_tcpv6;
```
these 2 lines should not be needed if you include `haproxy/proto_tcp.h`

matttbe commented on src/cfgparse-global.c:
```diff
@@ -24,6 +24,11 @@
 #include <haproxy/peers.h>
 #include <haproxy/protocol.h>
 #include <haproxy/tools.h>
+#include <haproxy/proto_rhttp.h>
```
it looks like they sort this by alphabetical order. Can you move it (and the tcp one, see below) under `haproxy/protocol.h`?

matttbe commented on src/cfgparse.c:
```diff
@@ -1192,8 +1192,12 @@ int cfg_parse_mailers(const char *file, int linenum, char **args, int kwm)
 			goto out;
 		}
-
+		
```
trailing whitespaces

matttbe commented on src/cfgparse.c:
```diff
@@ -1192,8 +1192,12 @@ int cfg_parse_mailers(const char *file, int linenum, char **args, int kwm)
+		if (proto->sock_prot != IPPROTO_TCP && proto->sock_prot != IPPROTO_MPTCP) {
+#else
 		if (proto->sock_prot != IPPROTO_TCP) {
+#endif
```
What about including `haproxy/proto_tcp.h`, removing the `ifdef` and just do:

```c
if (proto->sock_prot != proto_tcpv4.sock_prot) {
```

See my previous comment: https://github.com/CrapsDorian/haproxy/pull/1#discussion_r1587652459

(with the current solution, you would have to define `IPPROTO_MPTCP` if it was not defined)

matttbe reviewed (changes requested):
thx for the update. Some small comments, almost there

matttbe commented on src/cfgparse.c:
```diff
@@ -1192,8 +1193,7 @@ int cfg_parse_mailers(const char *file, int linenum, char **args, int kwm)
 			err_code |= ERR_ALERT | ERR_FATAL;
 			goto out;
 		}
-
```
do not modify this line (empty line before the `if`)

matttbe commented on reg-tests/tcp-rules/default_rules_mptcp.vtc:
```diff
@@ -0,0 +1,65 @@
+varnishtest "Test declaration of TCP rules in default sections"
```
```suggestion
varnishtest "Test declaration of TCP rules in default sections, with MPTCP"
```

CrapsDorian committed:
FEATURE: add opt-in MPTCP support

Multipath TCP (MPTCP), standardized in RFC8684 [1], is a TCP extension
that enables a TCP connection to use different paths.

Multipath TCP has been used for several use cases. On smartphones, MPTCP
enables seamless handovers between cellular and Wi-Fi networks while
preserving established connections. This use-case is what pushed Apple
to use MPTCP since 2013 in multiple applications [2]. On dual-stack
hosts, Multipath TCP enables the TCP connection to automatically use the
best performing path, either IPv4 or IPv6. If one path fails, MPTCP
automatically uses the other path.

To benefit from MPTCP, both the client and the server have to support
it. Multipath TCP is a backward-compatible TCP extension that is enabled
by default on recent Linux distributions (Debian, Ubuntu, Redhat, ...).
Multipath TCP is included in the Linux kernel since version 5.6 [3]. To
use it on Linux, an application must explicitly enable it when creating
the socket. No need to change anything else in the application.

This attached patch adds the "mptcp" global option in the config, which
allows the creation of an MPTCP socket instead of TCP on Linux. If
Multipath TCP is not supported on the system, an error will be reported,
and the application will stop.

A test has been added, it is a copy of "default_rules.vtc" in tcp-rules
with the addition of "mptcp" in the config. I'm not sure what else needs
to be tested for the moment, with this global MPTCP option.

Note: another patch is coming to support enabling MPTCP per address, but
I prefer to already send this patch, just in case, as I will soon have
less time to dedicate to this.

Due to the limited impact within a data center environment,
we have decided not to implement MPTCP between the proxy and the servers.
The high-speed, low-latency nature of data center networks reduces
the benefits of MPTCP, making the complexity of its implementation
unnecessary in this context.

Developed with the help of Matthieu Baerts (matttbe@kernel.org) and
Olivier Bonaventure (Olivier.Bonaventure@uclouvain.be)

Link: https://www.rfc-editor.org/rfc/rfc8684.html [1]
Link: https://www.tessares.net/apples-mptcp-story-so-far/ [2]
Link: https://www.mptcp.dev [3]
Signed-off-by: Matthieu Baerts (NGI0) <matttbe@kernel.org>
